services/yellow/src/api/Services/CreditCardPaymentService.cs

```diff
@@ -0,0 +1,152 @@
+        
+        public CreditCardPaymentService(IDataStoreClient dataStoreClient, bool failAllPayments)
+        {
+            _dataStoreClient = dataStoreClient;
+            _failAllPayments = failAllPayments;
+        }
+        
+        public CreditCardPaymentService(IDataStoreClient dataStoreClient)
+        {
+            _dataStoreClient = dataStoreClient;
+        }
+        
+        /// <inheritdoc/>
+        public async Task<AuthorizeResult> Authorize(decimal amount, CreditCard creditCard)
+        {
+            if (!CreditCardHasTheRequiredAmount(amount, creditCard))
+                return new AuthorizeResult(null, new InvalidOperationException("The given credit card has not the required amount"));
+
+            var paymentAuthorizationToken = await GetPaymentAuthorizationToken();
+
+            return new AuthorizeResult(paymentAuthorizationToken.Token, null);
+        }
+
+        /// <inheritdoc/>
+        public async Task<ChargeResult> Charge(decimal amount, CreditCard creditCard)
+        {
+            if (!CreditCardHasTheRequiredAmount(amount, creditCard))
+            {
+                return new ChargeResult()
+                {
+                    Exception = new InvalidOperationException("The given credit card has not the required amount")
+                };
+            }
+            
+            var paymentAuthorizationToken = await GetPaymentAuthorizationToken();
+
+            return await Charge(amount, paymentAuthorizationToken.Token!);
+        }
+
+        /// <inheritdoc/>
+        public async Task<ChargeResult> Charge(decimal amount, string authorizePaymentToken)
+        {
+            var (isValid, chargeResult) = await ValidateToken(authorizePaymentToken);
+
+            if (!isValid)
+                return chargeResult;
+
+            try
+            {
+                var paymentCode = await Pay(amount, authorizePaymentToken);
+                
+                var transaction = new Transaction(Guid.NewGuid().ToString(), paymentCode, amount, DateTime.Now);
+
+                await _dataStoreClient.InsertOneAsync(transaction);
+
+                return new ChargeResult
+                {
+                    transaction = transaction,
+                    AuthorizePaymentToken = authorizePaymentToken
+                };
+            }
+            catch (Exception e)
+            {
+                return new ChargeResult
+                {
+                    Exception = e
+                };
+            }
+        }
+
+        private async Task<(bool isValid, ChargeResult chargeResult)> ValidateToken(string authorizePaymentToken)
+        {
+            var t = await _dataStoreClient
+                .GetAsync<PaymentAuthorizationToken>(ExpressionCombinationOperator.And,
+                    token => token.Token == authorizePaymentToken);
+            
+            if (t is null)
+            {
+                return (false, new ChargeResult

```

Comment:
Instead of returning a `ChargeResult` from a `ValidateToken` method, have you consider throwing the exception directly?